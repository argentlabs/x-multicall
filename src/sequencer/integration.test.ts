import { describe, expect, mock, Mock, test } from "bun:test";
import { SequencerBatchProvider } from "./index";
import {
  Call,
  CallContractResponse,
  constants,
  SequencerProvider,
} from "starknet";
import { filterError } from "../utils.test";

interface MinimalMockProviderInterface {
  callContract: Mock<(request: Call) => Promise<CallContractResponse>>;
}

function getIntegrationProvider(): MinimalMockProviderInterface {
  const provider = new SequencerProvider({
    network: constants.NetworkName.SN_MAIN,
  });
  return {
    callContract: mock(async (call) => {
      const { contractAddress, entrypoint, calldata = [] } = call;
      const result = await provider.callContract({
        contractAddress,
        entrypoint,
        calldata,
      });
      return result;
    }),
  };
}

describe("SequencerBatchProvider", () => {
  test("should return the correct result for one call", async () => {
    const provider = getIntegrationProvider();
    const mc = new SequencerBatchProvider(provider);
    expect(provider.callContract.mock.calls.length).toEqual(0);

    const { result } = await mc.callContract({
      contractAddress:
        "0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7",
      entrypoint: "balanceOf",
      calldata: ["0xdeadbeef0"],
    });

    expect(result).toEqual(["0x0", "0x0"]);
    expect(provider.callContract.mock.calls.length).toEqual(1);
  });

  test("should return the correct result for multiple calls with one request", async () => {
    const provider = getIntegrationProvider();
    const mc = new SequencerBatchProvider(provider);

    const responses = await Promise.all(
      new Array(4).fill(null).map((_, i) =>
        mc.callContract({
          contractAddress: `0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7`,
          entrypoint: "balanceOf",
          calldata: [`0xdeadbeef${i}`],
        })
      )
    );

    expect(responses).toEqual([
      { result: ["0x0", "0x0"] },
      { result: ["0x0", "0x0"] },
      { result: ["0x0", "0x0"] },
      { result: ["0x0", "0x0"] },
    ]);
    expect(provider.callContract.mock.calls.length).toEqual(1);
  });

  test("one call fails in a batch", async () => {
    const provider = getIntegrationProvider();
    const mc = new SequencerBatchProvider(provider);

    const responses = await Promise.allSettled(
      new Array(4).fill(null).map((_, i) => {
        if (i === 2) {
          return mc.callContract({
            contractAddress: `0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7`,
            entrypoint: "balance_of",
            calldata: [`0xdeadbeef${i}`],
          });
        }
        return mc.callContract({
          contractAddress: `0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7`,
          entrypoint: "balanceOf",
          calldata: [`0xdeadbeef${i}`],
        });
      })
    );

    expect(filterError(responses)).toMatchSnapshot();
    expect(provider.callContract.mock.calls.length).toEqual(2);
  });

  test("two call fails in a batch", async () => {
    const provider = getIntegrationProvider();
    const mc = new SequencerBatchProvider(provider);

    const responses = await Promise.allSettled(
      new Array(4).fill(null).map((_, i) => {
        if (i === 0 || i === 2) {
          return mc.callContract({
            contractAddress: `0x053c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8`,
            entrypoint: "balance_of",
            calldata: [`0xdeadbeef${i}`],
          });
        }
        return mc.callContract({
          contractAddress: `0x053c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8`,
          entrypoint: "balanceOf",
          calldata: [`0xdeadbeef${i}`],
        });
      })
    );

    expect(filterError(responses)).toMatchSnapshot();
    expect(provider.callContract.mock.calls.length).toEqual(3);
  });
});
