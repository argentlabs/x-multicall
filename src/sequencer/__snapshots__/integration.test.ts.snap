// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`SequencerBatchProvider one call fails in a batch 1`] = `
[
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
  {
    "reason": [GatewayError: Error at pc=0:25:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:128)
Unknown location (pc=0:68)
Unknown location (pc=0:160)
Unknown location (pc=0:199)
Unknown location (pc=0:199)
Error message: multicall 2 failed
Unknown location (pc=0:177)

Error in the called contract (0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7):
Error at pc=0:104:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:1678)
Unknown location (pc=0:1664)

Error in the called contract (0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7):
Entry point 0x35a73cd311a05d46deda634c5ee045db92f811b4e74bca4437fcb5302b7af33 not found in contract with class hash 0x2760f25d5a4fb2bdde5f561fd0b44a3dee78c28903577d37d669939d97036a0.],
    "status": "rejected",
  },
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
]
`;

exports[`SequencerBatchProvider two call fails in a batch 1`] = `
[
  {
    "reason": [GatewayError: Error at pc=0:25:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:128)
Unknown location (pc=0:68)
Unknown location (pc=0:160)
Error message: multicall 0 failed
Unknown location (pc=0:177)

Error in the called contract (0x53c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8):
Error at pc=0:105:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:1683)
Unknown location (pc=0:1669)

Error in the called contract (0x53c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8):
Entry point 0x35a73cd311a05d46deda634c5ee045db92f811b4e74bca4437fcb5302b7af33 not found in contract with class hash 0x2760f25d5a4fb2bdde5f561fd0b44a3dee78c28903577d37d669939d97036a0.],
    "status": "rejected",
  },
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
  {
    "reason": [GatewayError: Error at pc=0:25:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:128)
Unknown location (pc=0:68)
Unknown location (pc=0:160)
Unknown location (pc=0:199)
Error message: multicall 1 failed
Unknown location (pc=0:177)

Error in the called contract (0x53c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8):
Error at pc=0:105:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:1683)
Unknown location (pc=0:1669)

Error in the called contract (0x53c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8):
Entry point 0x35a73cd311a05d46deda634c5ee045db92f811b4e74bca4437fcb5302b7af33 not found in contract with class hash 0x2760f25d5a4fb2bdde5f561fd0b44a3dee78c28903577d37d669939d97036a0.],
    "status": "rejected",
  },
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
]
`;
